### ISSUE TITLE
Add the ability to capture Standard Input

### ISSUE THREAD

**Issue by deanebarker** [statiqdev/Statiq#222]
This would make Wyam "pipe aware."  We could pass the contents of the initial document in on the command line call to the EXE.

We can create a module that returns a single document from STDIN, but that would have to be captured by the EXE.  I think it should be captured and assigned to some property on the ExecutionContext.  Just assign it and leave it sitting there.

Then, a module called `DocumentFromStdIn()` would read it off the ExecutionContext and turn it into a document.

All the pieces are there except the capturing of STDIN and assignment onto ExecutionContext, I think.

**Comment by daveaglick** [comment]
YES, SO MUCH THIS!

This lines up great with thoughts I've been having about better CI and cross platform support (which isn't coming soon, but will happen eventually). Turning Wyam into a stage in a larger OS-mediated pipeline opens up so many crazy doors. Like, I could envision grepping files for some keyword, sending them to Wyam via a pipe, and then having Wyam dump out a report.

**Comment by deanebarker** [comment]
One of the problems is that you have reserved the first parameter for the config file:

```
syntax.DefineParameter("root", ref _rootPath, DirectoryPath.FromString, "The folder (or config file) to use.")
```

So any standard input beyond the switch options goes to that.

Can we institute a breaking change whereby the first (and only) parameter is considered to be STDIN?  Root would need to be specified via an option, henceforth.

**Comment by daveaglick** [comment]
I had thought stdin was separate from the args you get passed in `Main()` - I.e., to read standard input you write `Console.ReadLine()` after processing arguments passed to main. So this should work?

```
echo "Foobar" | wyam.exe -p "c:\siteroot"
```

Should be easy enough to test...

**Comment by deanebarker** [comment]
You're right. I was thinking of something entirely different.

Working on it now.  I'll have a rough version shortly, which I know will need feedback and refinement, since I'm further into the internals than I'd rather be...

**Comment by deanebarker** [comment]
It works.  Here are the details, which we can hash through.

**Step 1: Get the STDIN into Program**

I added a private field to `Program`:

```
private string _stdin;
```

Then the trick is grabbing the STDIN without waiting for input.  I found some code [here](https://stackoverflow.com/questions/199528/c-sharp-console-receive-input-with-pipe):

```
var pipedText = "";
bool isKeyAvailable;

try
{
    isKeyAvailable = System.Console.KeyAvailable;
}
catch (InvalidOperationException expected)
{
    pipedText = System.Console.In.ReadToEnd();
}

_stdin = pipedText;
```

I don't know if there's a better way to do that, but `Console.ReadLine()` will sit and wait for input.

The code above worked, and it got the STDIN into the `_stdin` field.

**Step 2: Get the data into the Engine**

I created an `ApplicationInput` property on `Engine`.

I changed from "_stdin" to "ApplicationInput" here since the concept of "standard input" only makes sense in the context of a command line application.  If Wyam was ever executed some other way, perhaps that other thing would provide startup input differently (startup arguments for a Windows service, for example), so I figured it was better to get more generic once we moved away from the command line wrapper.

I set the `Engine.ApplicationInput` property just after the engine is instantiated (at the same time all the IO params are set, etc.)

**Step 3: Get the data into the ExecutionContext**

I created `ApplicationInput` properties on both `IExecutionContext` and `ExecutionContext`.  The latter simple passes through the property from `Engine`, much like a bunch of other properties on that class.

```
public string ApplicationInput => Engine.ApplicationInput;
```

**Step 4: Get the data into a Document via a module**

I created a one line `ReadApplicationInput` module:

```
return new IDocument[] { context.GetDocument(context.ApplicationInput) };
```

The config looks like this:

```
Pipelines.Add("StandardInputTest",
    ReadApplicationInput(),
    Meta("WritePath", "stdin.html"),
    WriteFiles()
);
```

And it works as expected.

To sum up, the data flows like this:

```
STDIN > Program > Engine > ExecutionContext > ReadApplicationInput > [BOOM, document!]
```

![](https://49.media.tumblr.com/ad72fd06dd438fcb38785080202def1b/tumblr_mwz5n31ZG61qb0snwo1_250.gif)

```
```

**Comment by daveaglick** [comment]
https://s-media-cache-ak0.pinimg.com/originals/b5/e9/46/b5e946fd3c44fb224930594d79dc4e32.gif

Seriously, this is great! I especially like the general mechanism of passing the input through to the engine and then making it available from the execution context for a module to pick up. Really A+.

Only part I'm not thrilled with is having to use an exception to check for available input. I may play around to see if there's an alternative (I seem to remember there was a way to replace the std stream readers or something). Don't let that stop you from submitting a PR when you think it's ready though - it's not a show-stopper and can be refined after the fact.

**Comment by daveaglick** [comment]
![image](https://cloud.githubusercontent.com/assets/1020407/13024627/b9b01f02-d1c5-11e5-9889-7f83bb0f8bd2.gif)

Seriously, this is great! I especially like the general mechanism of passing the input through to the engine and then making it available from the execution context for a module to pick up. Really A+.

Only part I'm not thrilled with is having to use an exception to check for available input. I may play around to see if there's an alternative (I seem to remember there was a way to replace the std stream readers or something). Don't let that stop you from submitting a PR when you think it's ready though - it's not a show-stopper and can be refined after the fact.

**Comment by deanebarker** [comment]
Yeah, I don't like that part either. Let's look for a better solution before I submit anything. I'll look around this weekend.

**Comment by daveaglick** [comment]
Some combination of `Console.SetIn()` and/or `Console.OpenStandardInput()` may do the trick. At least enough to Google/StackOverflow...

**Comment by deanebarker** [comment]
I might have found it.

```
if (Console.IsInputRedirected)
{
    using (var reader = new StreamReader(Console.OpenStandardInput(), Console.InputEncoding))
    {
        _stdin = reader.ReadToEnd();
    }
}
```

I'll do some more testing, but it seems to work (`IsInputRedirected` is new in .NET 4.5).

**Comment by daveaglick** [comment]
Resolved by #226
